### PR TITLE
Modifying semantics for GetSendMessage and GetSerializedSendMessage. Also adding ModifySendMessage

### PR DESCRIPTION
**Pull request by yashykt**
This PR makes the following interception API changes -
1) Serialization of message is now delayed for sync and callback APIs. Serialization is now only done after all interceptors are done running or when an interceptor calls GetSerializedSendMessage, whichever happens earlier.
2) Invoking GetSerializedSendMessage now has a side-effect that all future calls to GetSendMessage will return nullptr. 
3) New API method - ModifySendMessage allows interceptors to replace the message to be sent with a new message. Note that the interceptor is responsible for maintaining the life of the message for the duration of the Send message op, i.e., till POST_SEND_MESSAGE is received.

In most cases, interceptors will be prefering GetSendMessage() over GetSerializedSendMessage() if available.

### CODE REVIEW

File: include/grpcpp/impl/codegen/call_op_set.h [hunk 1]
```diff
@@ -352,22 +355,31 @@ class CallOpSendMessage {
   bool hijacked_ = false;
   ByteBuffer send_buf_;
   WriteOptions write_options_;
+  std::function<Status(const void*)> serializer_;
```
[the remark this answers is not in the document]
Comment (yashykt) [reply include/grpcpp/impl/codegen/call_op_set.h]: ah! I forgot about that

**Comment by vjpai** [comment]
Approved. I like the examples, but looks like there are still some leaks. I think it's just a matter of making sure that the function is nulled out on all paths, may be some trial and error there.

**Comment by yashykt** [comment]
That solved it! :)
Thanks for reviewing!

**Comment by tsalomie** [comment]
My current understanding of the outcome of this change is:
1) an interceptor can either access the request payload in serialized form or in original form, but not both at the same time.
2) you can go from original form to serialized form, but not back.
3) changing the request is possible in both original and serialized form.
4) changing the original form requires holding onto the new payload until POST_SEND_MESSAGE or until explicitly calling GetSerializedSendMessage()
5) changing the serialized form has not requirement.
6) with chained interceptors, once one moves to serialize the payload, that is final.

Are the above correct?

**Comment by vjpai** [comment]
@tsalomie Here's a partial response

1. Basically. If you access the serialized form, you're invalidating the non-serialized form since the serialized form is mutable.
1. As above.
1. Yes! The serialized form is mutable. The non-serialized form requires the ModifySendMessage call.
1. Yes. Changing the original form means that the interceptor is the owner of the new payload.
1. Correct. The library is always the owner of the serialized form.
1. Basically. You could deserialize the message again and use ModifySendMessage to re-establish a non-serialized form.

File: include/grpcpp/impl/codegen/call_op_set.h [hunk 2]
```diff
@@ -369,22 +378,32 @@ class CallOpSendMessage {
...
+        *static_cast<const M*>(message), send_buf_.bbuf_ptr(), &own_buf);
+    if (!own_buf) {
+      send_buf_.Duplicate();
+    }
+    return result;
+  };
+  // Serialize immediately only if we do not have access to the message pointer
+  if (msg_ == nullptr) {
+    return serializer_(&message);
+    serializer_ = nullptr;
```
Comment (vjpai) [inline include/grpcpp/impl/codegen/call_op_set.h]: oops! Surprising that this passed the tests.

Comment (vjpai) [reply include/grpcpp/impl/codegen/call_op_set.h]: Like I'm actually surprised that the compiler doesn't have an "unreachable code" warning.

Comment (yashykt) [reply include/grpcpp/impl/codegen/call_op_set.h]: yeah! I know. I would hope the compiler detects the non-reachable codepath.. or maybe we have just disabled that flag

**Comment by yashykt** [comment]
Thanks for your comment Tudor! I have stolen your points to add them as part of the documentation in https://github.com/grpc/grpc/pull/17662 :)

**Pull request merged**